### 1.	Install Python (if you don’t have it already), and install NLTK.  

In [1]:
import sys
import nltk

from nltk.book import *
from nltk.corpus import words
import re
import time
import requests

*** Introductory Examples for the NLTK Book ***
Loading text1, ..., text9 and sent1, ..., sent9
Type the name of the text or sentence to view it.
Type: 'texts()' or 'sents()' to list the materials.
text1: Moby Dick by Herman Melville 1851
text2: Sense and Sensibility by Jane Austen 1811
text3: The Book of Genesis
text4: Inaugural Address Corpus
text5: Chat Corpus
text6: Monty Python and the Holy Grail
text7: Wall Street Journal
text8: Personals Corpus
text9: The Man Who Was Thursday by G . K . Chesterton 1908


## 2.	Follow the instructions in chapter 1 of Bird-Klein for implementing a “lexical diversity” scoring routine.

In [22]:
start_line = '\*\*\* START OF THIS PROJECT GUTENBERG EBOOK.*\*\*\*'
end_line = '\*\*\* END OF THIS PROJECT GUTENBERG EBOOK.*\*\*\*'
print(start_line)

\*\*\* START OF THIS PROJECT GUTENBERG EBOOK.*\*\*\*


## 3. Go to http://www.gutenberg.org/wiki/Children%27s_Instructional_Books_(Bookshelf), and obtain three texts (of different grade levels) from the “Graded Readers” section. Report the lexical diversity score of each. Explain whether the result was surprising.

In [23]:
sixth = open('The_Ontario_Readers_Sixth_Book.txt', encoding="utf8").read()
third = open('The_Ontario_Readers_Third_Book.txt', encoding="utf8").read()
high = open('The_Ontario_Readers_The_High_School_Reader_1886.txt', encoding="utf8").read()

In [24]:
import re

# start line and end line for text
def remove_meta(text):
    return re.split(end_line, re.split(start_line, text)[1])[0]

from nltk.tokenize import word_tokenize

def vocab_count(text):
    if text is str:
        text = split_text(remove_meta(text))
    return len(set(text))
def split_text(text):
    return re.split("[^a-z']+")
#count the distinct tokens for each text
def tokens(text):
    text = nltk.word_tokenize(remove_meta(text))
    text = [word for word in text if word.isalpha()]
    return text 

In [25]:
def vocab_score(text):
    return len(set(tokens(text))) / len(text)

In [26]:
vocab_count(high)

103

In [27]:
def lexical_diversity(text):
    return len(set(tokens(text))) / len(tokens(text))

In [28]:
lexical_diversity(high)

0.12042005086553449

In [29]:
lexical_diversity(sixth)

0.1155387010380776

In [30]:
lexical_diversity(third)

0.13364178587727996

###	The following is the lexical diversity scores for each of the books I checked.
*   The Ontario Sixth Reader

>* 	lexical diversity = .1155 % 

*	The Ontario Third Reader

>*	lexical diversity = .1336 %

*	The Ontario High School Reader

>*	lexical diversity = .1204 %

*	You would think the High School Reader would have the highest lexical score but it was the 3rd graders. The actual lexical diversity results suggest there is no correlation between lexical diversity and the grader for each work since they are all very close in score. It is surprising the 3rd grade has the highest lexical score since you think they wouldn't be as diverse in their words.

## 4.	Also compare the vocabulary size of the same three texts. Explain whether the result was surprising.  

In [31]:
def distinct_tokens(text):
    return len(set(tokens(text)))
# make function to count vocab_size
def vocab_size(text):
    return len(tokens(text))

In [37]:
vocab_size(sixth)

135539

In [33]:
vocab_size(third)

63543

In [34]:
vocab_size(high)

121890

In [38]:


def vocab_score(text):
    return vocab_size(text) / len(text)

In [40]:
vocab_score(sixth)

0.16450145520658108

###### The results above are not suprising since a child's vocabulary size increases as they progress with grade levels. The third had the lowest vocab size but had the highest lexical score. It is not surprising that the higher grades had a bigger vocab size but it doesn't correlate to the lexical scores.  

## 5.	Write a paragraph arguing whether vocabulary size and lexical diversity in combination could be a better measure of text difficulty (or reading level) than either measure is by itself.

###### If a text has a higher index of lexical diversity it is likely to be more complex, more advanced and more difficult (if other things are equal). The main problem is that some measures of lexical diversity do not take account fully of differences in text length. In other words, if they are used on texts of different lengths they can give misleading results so maybe you would want to use a combination of both of them. 